### ISSUE TITLE
User Agent Randomization (Per-Session and/or Per-Site)

### ISSUE THREAD

**Issue by AzureB1te** [bromite/bromite#1529]
It's a feature request

Not realted to either. 
### Is your feature request related to privacy?

 It could be a massive boost to privacy, more precisely fingerprinting. 

### Is there a patch available for this feature somewhere?

I don't personally know of any 

### Describe the solution you would like

Bromite already implements a hard-coded UA. So what is that hard coded UA header changed, every time you restarted bromite? Either that, or for an even better (but probabily much harder) way, a different UA header for every site, the same way each sites gets it's random time-zone. Even better, if both get impemented.
This, at lease based on my understanding would massively increase privacy, with a minimal impact to user experience. 

### Describe alternatives you have considered

N/A

**Comment by uazo** [comment]
> This, at lease based on my understanding would massively increase privacy, with a minimal impact to user experience.

I take advantage of this request to better understand something.
I always thought that to reduce fingerprinting, the technique was to hide in the crowd, but browsers adopt the technique of making themselves unique, all except tor, from what I understand.
I tried to read up on it but the articles are full of statistical formulas, entropy and bits, and I didn't quite understand the meaning.
based on your request I guess you have documented it, could you give me some explanation about it?

**Comment by AzureB1te** [comment]
> > This, at lease based on my understanding would massively increase privacy, with a minimal impact to user experience.
> 
> I take advantage of this request to better understand something. I always thought that to reduce fingerprinting, the technique was to hide in the crowd, but browsers adopt the technique of making themselves unique, all except tor, from what I understand. I tried to read up on it but the articles are full of statistical formulas, entropy and bits, and I didn't quite understand the meaning. based on your request I guess you have documented it, could you give me some explanation about it?

The concept of fingerprint randomization is the follwong : 
Modern browsers make it really hard to hide in the crowd. For example, this is almost impossible with a fork-browser like Bromite, or for that sake Brave, unless there are a lot of people using that browser. 
So the idea of randomization is to instead of trying to blend in, you change certain parts of the UA, canvas, audio, webgl, fonts, etc. every time the browser (or a new tab) is opened. This way, if you visit a website with a randomized fingerprint, it will have a much harder time, to link that back to you, the next time you visit. Now there's obviously more things at play here, like your IP adress and cookies, but implementing this UA Randomization feature would be a massive improvement as I've already said. 
So basically, you're not trying to hide the fact that you're unique but you change what's unique about you, so it can't be tied to you. 
Hope I did a good explanation

**Comment by uazo** [comment]
> Modern browsers make it really hard to hide in the crowd. For example, this is almost impossible with a fork-browser like Bromite, or for that sake Brave, unless there are a lot of people using that browser.

i'm not sure about this...
There is a project in chromium called privacy budget which is now in the data acquisition phase, for which they are tracing the sources of possible fingerprinting directly from the code, and this makes me think that we may know what the tracing data may be.
I don't know if that data will ever be in the public domain, but it seems to me that the road is towards flattening the differences, even if I don't know for sure

> So the idea of randomization is to instead of trying to blend in

maybe because it's easier?

> Hope I did a good explanation

sure, and I thank you

**Comment by AzureB1te** [comment]
> maybe because it's easier?
Welp, Brave uses the same methodology. Firefox's privacy.resistFingerprinting also does this, along with other things. I'm not sure if it being easier is the reason for this, but as far as I can see, it's gaining popularity. I'm no developer (yet, still in school), but I think my feature request would "fit" Bromite's featureset.

**Comment by uazo** [comment]
> but I think my feature request would "fit" Bromite's featureset.

I don't have the elements to judge.
however, for example, the timezone override should be used to mirror the real location of your ip, if for example you are using a vpn. but if everyone else using that ip doesn't, it makes you more identifiable.
similarly overriding the timezone as random is more of a game than a function to use, because it may make you more identifiable.
so too, tweaking the ua to look like firefox, for example, or windows, might make you more identifiable because surely there is a way to check if you are lying.
another thing is if everyone does it, but if it's just you ...

**Comment by csagan5** [comment]
> this is almost impossible with a fork-browser like Bromite

Bromite does not try to hide in the crowd, that is not possible.

> I don't know if that data will ever be in the public domain, but it seems to me that the road is towards flattening the differences, even if I don't know for sure

This is what Vanadium attempts to do; as for upstream Chromium: I will believe it when I see it.

As for UA randomization: it will lead to the occasional broken site and reproducing the issue will be very difficult. I do not think it is worth adding this to Bromite.

**Comment by AzureB1te** [comment]
> As for UA randomization: it will lead to the occasional broken site and reproducing the issue will be very difficult. I do not think it is worth adding this to Bromite.
>What if randomization would only happen on breakage-safe parameters (for exanple device model, android version)? Woulen't that work?

**Comment by csagan5** [comment]
The benefit of this feature is zero or too little to justify the effort and complexity needed to maintain it.
